### DISCUSSION TITLE
Why does Micronaut return invalid, plain JSON strings?

### DISCUSSION THREAD

**Discussion by saw303** [micronaut-projects/micronaut-core#11118] [Q&A] [answered]
When returning a plain String within a Micronaut controller such as

```java
@Get ("/")
public HttpResponse<?> hello() {
    return HttpResponse.notFound("Hello");
}
```

the controller returns 

```
HTTP/1.1 404 Not Found
date: Mon, 26 Aug 2024 17:40:05 GMT
content-type: application/json
content-length: 5

Hello
```

or with this one

```java
@Get("/hello")
public String hello2() {
    return "Hello";
}
```

the controller returns 

```
HTTP/1.1 200 OK
date: Mon, 26 Aug 2024 17:40:55 GMT
content-type: application/json
content-length: 5

Hello
```

Syntactically `Hello` is not a valid JSON string, but `"Hello"` is. Why is Micronaut not returning a proper JSON string value with this implementation, but declares the response to be `application/json`? Is this a bug or am I missing something?

If I use an `ObjectMapper` to serialise `Hello` into a JSON string, everything works as expected and a proper JSON string is returned to the client.

```java
@Get("hello3")
public String hello3() throws IOException {
    return objectMapper.writeValueAsString("Hello");
}
```

results in

```
HTTP/1.1 200 OK
date: Mon, 26 Aug 2024 17:44:06 GMT
content-type: application/json
content-length: 7

"Hello"
```

**Comment by dstepanov** [comment, marked as answer]
The string is meant to be a raw type representing already encoded JSON

**Reply by saw303** [reply]
I understand. I stumpled over it, because it the first time we return a plain string there. When providing an object instance, the proper JSON serialization kicks in.

```java
@Get("hello4")
public HttpResponse<Hello> hello4() throws IOException {
    return HttpResponse.notFound(new Hello("Hello"));
}

@Serdeable
public record Hello(String name) {}
```

This results in 

```json
{
  "name": "Hello"
}
```

Is this inconsistent behaviour intentional?

**Reply by yawkat** [reply]
Yes. Returning string directly is considered a "raw" response, while other types are serialized.

**Reply by saw303** [reply]
Ok. Thank you for that clarification.
